File: 2021/10/python/main.py
points = {
    ')': 3,
    ']': 57,
    '}': 1197,
    '>': 25137
}

opening = "([{<"
closing = ")]}>"

def parse_line(line):
    stack = []

    for c in line:
        if c in opening:
            stack.append(c)
        else:
            if c != closing[opening.index(stack.pop())]:
                return (points[c], 0)
                
    score = 0
    while stack:
        score = score * 5 + opening.index(stack.pop()) + 1 

    return (0, score)

corrupt_score = 0
incomplete_scores = []
with open('../input.txt') as f:
    for line in f.readlines():
        c, i = parse_line(line.strip())
        corrupt_score += c
        if i:
            incomplete_scores.append(i)


print("Part 1:", corrupt_score)
print("Part 2:", sorted(incomplete_scores)[len(incomplete_scores)//2])

File: 2021/10/rust/src/main.rs
use phf::phf_map;
use std::error::Error;
use std::fs::File;
use std::io::{BufRead, BufReader};

static POINTS: phf::Map<char, u64> = phf_map! {
    ')' => 3,
    ']' => 57,
    '}' => 1197,
    '>' => 25137,
};

static OPENING: &str = "([{<";
static CLOSING: &str = ")]}>";

enum ParseResult {
    Corrupt(u64),
    Incomplete(u64),
}

fn parse(line: &str) -> ParseResult {
    let mut stack: Vec<char> = Vec::new();

    for c in line.chars() {
        if OPENING.contains(c) {
            stack.push(c);
        } else {
            let matching = stack.pop().unwrap();

            if OPENING.find(matching).unwrap() != CLOSING.find(c).unwrap() {
                return ParseResult::Corrupt(*POINTS.get(&c).unwrap());
            }
        }
    }

    let mut score = 0;
    while !stack.is_empty() {
        let c = stack.pop().unwrap();
        score = score * 5 + OPENING.find(c).unwrap() + 1;
    }

    ParseResult::Incomplete(score.try_into().unwrap())
}

fn main() -> Result<(), Box<dyn Error>> {
    let file = File::open("../input.txt")?;
    let reader = BufReader::new(file);

    let mut corrupt_score: u64 = 0;
    let mut incomplete_scores: Vec<u64> = Vec::new();

    for line in reader.lines() {
        match parse(line?.trim()) {
            ParseResult::Corrupt(n) => corrupt_score += n,
            ParseResult::Incomplete(n) => incomplete_scores.push(n),
        }
    }

    println!("Part 1: {}", corrupt_score);

    incomplete_scores.sort();
    println!("Part 2: {}", incomplete_scores[incomplete_scores.len() / 2]);

    Ok(())
}
